core/src/main/scala/chisel3/package.scala

```diff
@@ -234,6 +235,117 @@ package object chisel3 {
+     * 2. For bits - if no format specifier given (%s i.e) call .toPrintable to pick the default one. For regular data - let the String.format handle canonical case of %s too.
+     * Still testing this code - I am pretty sure this can be refactored and made more efficient / readable. WIP.
+     *
+     * /
+     */
+     */
+    def cf(args: Any*): Printable = {
+      sc.checkLengths(args) // Enforce sc.parts.size == pargs.size + 1
+      val parts = sc.parts.map(StringContext.treatEscapes)
+
+      // The 1st part is assumed never to contain a format specifier.
+      // If the 1st part of a string is an argument - then the 1st part will be an empty String.
+      // So we need to parse parts following the 1st one to get the format specifiers if any
+      val partsAfterFirst = parts.slice(1, parts.size)
+
+      // Align parts to their potential specifiers
+      val partsAndSpecifierSeq = partsAfterFirst.zip(args).map {
+        case (part, arg) => {
+
+          // Check if part starts with a format specifier (with % - disambiguate with literal % checking the next character if needed to be %)
+          // In the case of %f specifier there is a chance that we need more information - so capture till the 1st letter (a-zA-Z).
+          // Example cf"This is $val%2.2f here" - parts - Seq("This is ","%2.2f here") - the format specifier here is %2.2f.
+          val idx_of_fmt_str =
+            if (
+              !part.isEmpty() && part.charAt(0) == '%' && (part.size == 1 || (part.size >= 2 && part.charAt(1) != '%'))
+            ) part.indexWhere { _.isLetter }
+            else -1
+
+          // If no format specifier - pick default - %s
+
+          val fmt = if (idx_of_fmt_str >= 0) part.substring(0, idx_of_fmt_str + 1) else "%s"
+          val fmtArgs: Printable = arg match {
+            case b: Bits => {
+              require(fmt.size == 2, "In the case of bits, only single format char allowed!")
+              fmt match {
+                case "%s" => b.toPrintable
+                case "%n" => Name(b)
+                case "%N" => FullName(b)
+                // Default - let FirrtlFormat check validity of the format string to avoid repeating checks.
+                case f => FirrtlFormat(f.substring(1, 2), b)
+              }
+            }
+            case d: Data => {
+              require(
+                fmt == "%s" || fmt == "%n" || fmt == "%N",
+                "Non-bits only  allowed with (%s,%n, %N) format specifiers!"
+              )
+              fmt match {
+                case "%n" => Name(d)
+                case "%N" => FullName(d)
+                case "%s" => d.toPrintable
+                case x => {
+                  val msg = s"Illegal format specifier '$x'!\n"
+                  throw new UnknownFormatConversionException(msg)
+                }
+              }
+            }
```

Comment:
```suggestion
            case d: Data => Printable.pack(fmt, d)
```
Please correct me if I'm wrong, but I think `Printable.pack` already handles everything here (with the exception of `%s` which I think should be implemented as `.toString`.

Reply:
The only caveat is that delegating individual pieces to `Printable.pack` will give a better error message than this PR currently does, but a worse message than it could. A more ideal version of this PR would be to have `Printable.pack` delegate to this function (ie. I'm suggesting moving much of the `Printable.pack` logic here (including it's nice error messages with carats) and then changing `Printable.pack` to just construct a `StringContext` and call this method.

Reply:
@jackkoenig  my latest commit does this - for now named it packCF - also a known limitation is handling %% - WIP.... do take a look and give feedback!

Reply:
Also due to the fact that escaping % needs to be explicitly done by user for "cf" API - that part is not needed to be done explicitly when emittting firrtl.

Reply:
Delegated Printable.pack to use the cf API - retained all of the error reporting with the previous version and went one step ahead - also report unescaped literal % if found.